### PR TITLE
Fixes around namespaces and Fluent API calls

### LINKED ISSUE #23848: Generate fully-qualified Fluent API calls in model snapshot
This would be the proper fix for #23456 (which we temporarily worked around for 5.0).

### LINKED ISSUE #25046: Custom namespaces in scaffolding
When scaffolding, it's not possible include a Fluent API fragment that isn't in [the 3 hard-coded standard namesapces](https://github.com/dotnet/efcore/blob/main/src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs#L80). This forces all provider Fluent APIs to be defined under Microsoft.EntityFrameworkCore.Metadata.

I'm not so sure this is a bad thing - it's good for Fluent APIs to just light up without requiring special namespaces. However, this was the root cause of https://github.com/npgsql/efcore.pg/issues/1613, which made me move some stuff from Npgsql.EntityFrameworkCore.PostgreSQL.Metadata to Microsoft.EntityFrameworkCore.Metadata, and causing migrations issues as in #24877.

Our options:

1. Decide to support custom namespaces in scaffolding (this issue), and leave the Npgsql types in Npgsql.EntityFrameworkCore.PostgreSQL.Metadata. This could be a lot of work.
2. Decide not to support custom namespaces in scaffolding (all Fluent API stuff should be in Microsoft.EntityFrameworkCore.Metadata), and leave things as they are. PostgreSQL users upgrading from EF 5.0 to 6.0 need to manually update their migration files, model snapshots, etc.
3. Not support custom namespaces in scaffolding, but still revert the Npgsql namespace change. No more migration trouble for upgrading from 5.0 to 6.0, but scaffolding would require an extra gesture each time.

I personally vote for option 2, but let's discuss.

Could be related to #23848, which is about generating full-qualified (non-extension) Fluent API calls in the model snapshot. That issue would probably evolve MethodCallFragment into something more than a string, which could also help us generate the appropriate non-standard namespace too, if we want to do that.

### LINKED ISSUE #25434: Azure edition options annotations don't get generated as Fluent APIs
We don't scaffold these, but it's still better to have the Fluent API calls in the model snapshot, rather than raw annotations.

Will do this as part of the work on #23848

### PR DESCRIPTION
**Pull request by roji**
@bricelam this is a draft for getting your general OK before going further... #25046 (custom namespaces in scaffolding) is fully implemented, while #23848 (type-qualified Fluent APIs in the model snapshot) is a bit more tricky/ugly.

Closes #25046
Closes #23848
Closes #25434

### CODE REVIEW

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs [hunk 1]
```diff
@@ -1000,6 +1000,44 @@ public virtual string Fragment(MethodCallCodeFragment fragment)
             return builder.ToString();
         }
 
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual string TypeQualifiedMethodCall(MethodCallCodeFragment fragment, string instanceIdentifier)
```
Comment (roji) [inline src/EFCore.Design/Design/Internal/CSharpHelper.cs:1009]: This is, in a word, an awful hack and I hate it - but we need some way to pass in the first argument (i.e. the builder), as an identifier (where the other arguments are generated as literals). The separation of concerns between the CSharpHelper and CSharpSnapshotGenerator starts to break down here... Let me know if you have any other ideas.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs [hunk 2]
```diff
@@ -35,6 +37,70 @@ public class AnnotationCodeGenerator : IAnnotationCodeGenerator
             RelationalAnnotationNames.RelationalOverrides
         };
 
+        #region MethodInfos
+
+        private static readonly MethodInfo _hasDefaultSchemaMethodInfo
```
Comment (roji) [inline src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:42]: Now code generation can feel like the query pipeline!

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1009]: I'd prefer a `Fragment` overload with identifier and a `bool` for whether the call should be type qualified.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs [hunk 3]
```diff
@@ -1000,6 +1000,44 @@ public virtual string Fragment(MethodCallCodeFragment fragment)
...
+        {
+            // TODO: Strings
+            if (fragment.MethodInfo is null)
+            {
+                throw new ArgumentException("Cannot generate fully-qualified method calls for fragments without a MethodInfo", nameof(fragment));
+            }
+            if (fragment.ChainedCall is not null)
+            {
+                throw new ArgumentException("Cannot generate fully-qualified method calls for chained fragments", nameof(fragment));
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Design/Internal/CSharpHelper.cs:1018]: Type-qualified calls can have chained calls. But a static method call cannot be a chained call.

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 4]
```diff
@@ -58,41 +79,55 @@ public virtual void Generate(string builderName, IModel model, IndentedStringBui
...
                 stringBuilder.AppendLine(";");
+
+                if (fullyQualifiedFluentApiCalls.Count > 0)
+                {
+                    stringBuilder.AppendLine();
+                }
+            }
+
+            // Then generate fully-qualified calls
+            foreach (var call in fullyQualifiedFluentApiCalls)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:126]: Instead of splitting out type-qualified calls we should make all calls non-chained unless they are specified using `ChainedCall`. This is safer in general, as we don't check the return type.

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs [hunk 5]
```diff
@@ -110,7 +112,23 @@ public class CSharpDbContextGenerator : ICSharpDbContextGenerator
                 _sb.AppendLine("}");
             }
 
-            return _sb.ToString();
+            var namespaceStringBuilder = new StringBuilder();
+
+            var namespaces = _hardcodedNamespaces.Concat(_customNamespaces.OrderBy(ns => ns));
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:117]: Add `_hardcodedNamespaces` to `_customNamespaces` instead, remove `finalContextNamespace` from them and then split out `"System*"` before ordering. This way you can remove the `Conains` check when adding custom namespaces.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs [hunk 6]
```diff
@@ -218,9 +273,12 @@ public virtual void RemoveAnnotationsHandledByConventions(IIndex index, IDiction
         {
             var methodCallCodeFragments = new List<MethodCallCodeFragment>();
 
+            // TODO: Could be problematic... Which HasConstraintName needs to be invoked here (ReferenceReference, ReferenceCollection?)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:276]: Depends on whether the foreign key is unique. Also rename `navigation` to `foreignKey`.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs [hunk 7]
```diff
@@ -676,7 +729,26 @@ private static bool TryGetAndRemove<T>(IDictionary<string, IAnnotation> annotati
...
+                annotations.Remove(annotationName);
+                if (annotation.Value is object annotationValue)
+                {
+                    methodCallCodeFragments.Add(
+                        new MethodCallCodeFragment(methodInfo, annotationValue));
+                }
+            }
+        }
+
+        // TODO: Temporary
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:746]: I think the Obsolete attribute is enough

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs [hunk 8]
```diff
@@ -812,5 +850,13 @@ private void GenerateSequence(ISequence sequence)
         private IList<string> GenerateAnnotations(IEnumerable<IAnnotation> annotations)
             => annotations.Select(
                 a => $".HasAnnotation({_code.Literal(a.Name)}, {_code.UnknownLiteral(a.Value)})").ToList();
+
+        private void AddNamespace(string? ns)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:854]: We should also allow custom namespaces for literals. Perhaps pass in the `HashSet` as an argument in `CSharpHelper` calls

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1009]: @AndriySvyryd the problem is that when the call shouldn't be type-qualified, the identifier is generated externally by the caller, since multiple fragments are chained. In other words, when the type-qualified bool is false, identifier would never be used - this is why I went with two different methods.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1018]: I currently only generate type-qualified calls for static methods (see https://github.com/dotnet/efcore/pull/25435/files#diff-7b0d72a78a174967c02d622f305213f4e35f61da02481b5cd870c4ce7fa700c7R85); for instance methods, there's no potential ambiguity (the compiler always prefers instance methods to extension methods) so I'm not sure why we'd do type-qualified calls. Am I missing something?

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 4 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:126]: Would you also want us to do this for scaffolding? If so, that would make that code uglier and not in-line with how we usually document to use the Fluent APIs. If not, then whatever error we'd be avoiding in the model snapshot would appear there (in other words, seems like we should either go type-quailfied everywhere or nowhere?).

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 5 above)
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:117]: For perf? :)

The problem is that we want to order the hardcoded namespaces in a specific way, i.e. System before Microsoft.EntityFrameworkCore, although alphabetically it's the inverse. So we'd have specific code dealing with that, this seemed simpler. But am OK to do it your way, let me know.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs (hunk 6 above)
Comment (roji) [reply src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:276]: > Depends on whether the foreign key is unique

Yeah, that's stuff we didn't need to deal with before this...

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 8 above)
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:854]: Good point.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs (hunk 7 above)
Comment (roji) [reply src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:746]: The whole thing is only there temporarily while I'm working - as it's private I'll just remove it.

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 5 above)
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:117]: Or I can just remove _hardcodedNamespaces from _customNamespaces, and then output them one after the other, no need for a check. Will do that.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs (hunk 6 above)
Comment (roji) [reply src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:276]: One thought is to continue allowing creating MethodCallCodeFragment in the old way, without a MethodInfo (i.e. not obsolete it). This would allow using it in places where we assume no disambiguation is necessary, and possibly in places where we don't have access to the MethodInfo of the target method (though we currently don't think such cases exist).

Comment (AndriySvyryd) [reply src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:276]: Allowing calls without `MethodInfo` limits our ability to refactor the generators. Unless we find an important case where this is needed we should discourage it.

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 5 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:117]: `_customNamespaces` could also contain System namespaces.

In fact, if we allow adding namespaces for literals we can remove "System" from `_hardcodedNamespaces`

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:126]: It's just a refactoring to clarify the semantics. We can still output the same code, we'd just need to preprocess the fragments and link the compatible ones together explicitly using `ChainedCall` first.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1018]: > Am I missing something?

This check adds unnecessary restriction that type-qualified calls can't be chained on.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1009]: > In other words, when the type-qualified bool is false, identifier would never be used - this is why I went with two different methods.

What I am proposing is that when type-qualified bool is false and identifier is not null this should generate `identifier.Method()`

Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1009]: We can do this, but would it ever be used? Wouldn't we always prefer to chain multiple calls on the same builder (identifier), meaning that the identifier would have to come from outside CSharpHelper (like today)?

I assume we'd also need to throw if the type-quailfied flag is false and the identifier is null (since at least at the moment, MethodCallCodeFragment always represents either an instance call or an extension method call, both requiring an identifier).

If so, that means we'd use only 2 out of 4 possibilities:

TypeQualified? | Identifier null? | Generated result                                  | Usable?
-------------- | ---------------- | ------------------------------------------------- | -------
false          | true             | `.SomeMethod("arg1", "arg2")`                     | Yes
true           | false            | `SomeType.SomeMethod(identifier, "arg1", "arg2")` | Yes
false          | false            | `identifier.SomeMethod()`                         | Not sure when
true           | true             | `SomeType.SomeMethod(???, "arg1", "arg2")`        | Never, invalid

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1018]: This seems related to the discussion above (https://github.com/dotnet/efcore/pull/25435#discussion_r684764627)... As the code currently is, this new method is only used for static methods, whereas instance methods go through the existing overload which supports chaining.

If we indeed change to having a single method for all method call fragment types, then indeed we'd need to support chaining there for instance methods.

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 4 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:126]: Understood, will take a look.

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 5 above)
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:117]: Yep, agreed.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1018]: Will look into this.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1009]: Will play around with this in any case.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs (hunk 6 above)
Comment (roji) [reply src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:276]: Done, leaving unresolved for you to confirm.

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 8 above)
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs]: Split out to #25493

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs]: I've implemented chained, type-qualified calls too:

```c#
SqlServerModelBuilderExtensions
    .UseIdentityColumns(modelBuilder, 1L, 1)
    .HasDatabaseMaxSize(""100 MB"")
    .HasServiceTierSql(""'basic'"")
    .HasPerformanceLevelSql(""'S0'"");"),
```

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs [hunk 9]
```diff
@@ -967,15 +971,69 @@ private bool HandleList(IEnumerable<Expression> argumentExpressions, StringBuild
...
-        public virtual string Fragment(MethodCallCodeFragment fragment)
-            => Fragment(fragment, indent: 0);
+        public virtual string Fragment(MethodCallCodeFragment fragment, string? instanceIdentifier = null, bool typeQualified = false)
+            => Fragment(fragment, typeQualified, instanceIdentifier, indent: 0);
 
-        private  string Fragment(MethodCallCodeFragment fragment, int indent)
+        private string Fragment(MethodCallCodeFragment fragment, bool typeQualified, string? instanceIdentifier, int indent)
         {
-            var builder = new IndentedStringBuilder();
+            IndentedStringBuilder builder = new();
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Design/Internal/CSharpHelper.cs:979]: :trollface:

File: src/EFCore.Design/Properties/DesignStrings.Designer.cs [hunk 10]
```diff
@@ -63,6 +63,12 @@ public static string CannotFindTypeMappingForColumn(object? columnName, object?
                 GetString("CannotFindTypeMappingForColumn", nameof(columnName), nameof(dateType)),
                 columnName, dateType);
 
+        /// <summary>
+        ///     A type-qualified method call requires an instance identifier and a MethodInfo.
+        /// </summary>
+        public static string CannotGenerateTypeQualifiedMethodCal
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Properties/DesignStrings.Designer.cs:69]:
```suggestion
        public static string CannotGenerateTypeQualifiedMethodCall
```

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs [hunk 11]
```diff
@@ -967,15 +971,69 @@ private bool HandleList(IEnumerable<Expression> argumentExpressions, StringBuild
...
+                if (instanceIdentifier is not null)
+                {
+                    builder.Append(instanceIdentifier);
+
+                    if (current.ChainedCall is not null)
+                    {
+                        builder
+                            .AppendLine()
+                            .IncrementIndent();
+                    }
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Design/Internal/CSharpHelper.cs:1032]: 👍

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs [hunk 12]
```diff
@@ -989,30 +1047,40 @@ private  string Fragment(MethodCallCodeFragment fragment, int indent)
...
+                    break;
+                }
+
+                builder.AppendLine();
                 current = current.ChainedCall;
             }
 
             return builder.ToString();
+
+            void Arg(object? argument)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Design/Internal/CSharpHelper.cs:1066]: Spell it out pls

Comment (AndriySvyryd) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1066]: Also make it static or inline

Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1066]: Even local methods huh

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 13]
```diff
@@ -1621,23 +1515,91 @@ protected virtual void GenerateKeyAnnotations(IKey key, IndentedStringBuilder st
...
+
+                stringBuilder
+                    .AppendLines(Code.Fragment(typeQualifiedCalls, builderName, typeQualified: true), skipFinalNewline: true)
+                    .AppendLine(";");
+
+                // foreach (var call in typeQualifiedCalls)
+                // {
+                //     stringBuilder.Append(Code.Fragment(call, builderName, typeQualified: true));
+                //     stringBuilder.AppendLine(";");
+                // }
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:1601]: ❓

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 14]
```diff
@@ -459,94 +437,93 @@ public virtual void Generate(string builderName, IModel model, IndentedStringBui
...
 
-            using (stringBuilder.Indent())
-            {
-                if (property.IsConcurrencyToken)
-                {
-                    stringBuilder
-                        .AppendLine()
-                        .Append(".IsConcurrencyToken()");
-                }
+            stringBuilder.IncrementIndent();
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:482]: Where is the `DecrementIndent`?

Comment (AndriySvyryd) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:482]: Probably makes sense to put a comment here on similar places

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 15]
```diff
@@ -1621,23 +1515,91 @@ protected virtual void GenerateKeyAnnotations(IKey key, IndentedStringBuilder st
                 .Append(")");
         }
 
-        private static IReadOnlyList<IAnnotation> RemoveAmbiguousFluentApiAnnotations(
+        private void GenerateRemainingAnnotations(
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:1518]: Rename to GenerateAnnotations, signature is enough to describe what it does

File: src/EFCore/Design/MethodCallCodeFragment.cs [hunk 16]
```diff
@@ -63,12 +136,22 @@ public MethodCallCodeFragment(string method, params object?[] arguments)
...
         public virtual MethodCallCodeFragment? ChainedCall { get; }
 
+        /// <summary>
+        ///     Creates a method chain from this method to another.
+        /// </summary>
+        /// <param name="methodInfo"> The method's <see cref="MethodInfo" />. </param>
+        /// <param name="arguments"> The next method call's arguments. </param>
+        /// <returns> A new fragment representing the method chain. </returns>
+        public virtual MethodCallCodeFragment Chain(MethodInfo methodInfo, params object[] arguments)
+            => Chain(new MethodCallCodeFragment(methodInfo, arguments));
```
Comment (AndriySvyryd) [inline src/EFCore/Design/MethodCallCodeFragment.cs:146]: Would be nice to add a check that the types are compatible, but can be done in a separate PR

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 17]
```diff
@@ -1421,7 +1315,7 @@ protected virtual void GenerateKeyAnnotations(IKey key, IndentedStringBuilder st
             Check.NotNull(annotations, nameof(annotations));
             Check.NotNull(stringBuilder, nameof(stringBuilder));
 
-            foreach (var annotation in annotations)
+            foreach (var annotation in annotations.OrderBy(a => a.Name))
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:1318]: Why do you need to reorder?

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 18]
```diff
@@ -1621,23 +1515,91 @@ protected virtual void GenerateKeyAnnotations(IKey key, IndentedStringBuilder st
...
+
+            // Chain together all Fluent API calls which we can, and leave the others to be generated as type-qualified
+            foreach (var call in fluentApiCalls)
+            {
+                if (call.MethodInfo is not null
+                    && call.MethodInfo.IsStatic
+                    && (call.MethodInfo.DeclaringType is null
+                        || call.MethodInfo.DeclaringType.Assembly != typeof(RelationalModelBuilderExtensions).Assembly))
+                {
+                    typeQualifiedCalls = typeQualifiedCalls is null ? call : typeQualifiedCalls.Chain(call);
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:1539]: Why is it safe to chain calls that should be type-qualified?

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs [hunk 19]
```diff
@@ -28,7 +30,8 @@ public class CSharpDbContextGenerator : ICSharpDbContextGenerator
         private readonly ICSharpHelper _code;
         private readonly IProviderConfigurationCodeGenerator _providerConfigurationCodeGenerator;
         private readonly IAnnotationCodeGenerator _annotationCodeGenerator;
-        private IndentedStringBuilder _sb = null!;
+        private readonly IndentedStringBuilder _sb = new();
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:33]: Might as well rename this

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs [hunk 20]
```diff
@@ -110,7 +109,31 @@ public class CSharpDbContextGenerator : ICSharpDbContextGenerator
...
+                        _ => 4
+                    })
+                .ThenBy(ns => ns);
+
+            if (finalContextNamespace != modelNamespace && !string.IsNullOrEmpty(modelNamespace))
+            {
+                namespaces = namespaces.Append(modelNamespace);
+            }
+
+            foreach (var ns in namespaces)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:129]: @namespace

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs [hunk 21]
```diff
@@ -809,8 +820,48 @@ private void GenerateSequence(ISequence sequence)
...
+            foreach (var call in _annotationCodeGenerator.GenerateFluentApiCalls(annotatable, annotations))
+            {
+                var fluentApiCall = call;
+
+                // Remove optional arguments
+                if (fluentApiCall.MethodInfo is { } methodInfo)
+                {
+                    var methodParameters = methodInfo.GetParameters();
+
+                    for (int i = fluentApiCall.Arguments.Count - 1, j = methodInfo.IsStatic ? i + 1 : i; i >= 0; i--, j--)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:834]: You can calculate the offset outside and reduce fanciness

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs [hunk 22]
```diff
@@ -35,6 +37,78 @@ public class AnnotationCodeGenerator : IAnnotationCodeGenerator
...
+
+        private static readonly MethodInfo _propertyHasComputedColumnSqlMethodInfo2
+            = typeof(RelationalPropertyBuilderExtensions).GetRequiredRuntimeMethod(
+                nameof(RelationalPropertyBuilderExtensions.HasComputedColumnSql), typeof(PropertyBuilder), typeof(string));
+
+        private static readonly MethodInfo _hasComputedColumnSqlMethodInfo3
+            = typeof(RelationalPropertyBuilderExtensions).GetRequiredRuntimeMethod(
+                nameof(RelationalPropertyBuilderExtensions.HasComputedColumnSql), typeof(PropertyBuilder), typeof(string), typeof(bool));
+
+        private static readonly MethodInfo _propertySsFixedLengthMethodInfo
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:78]:
```suggestion
        private static readonly MethodInfo _propertyIsFixedLengthMethodInfo
```

File: src/EFCore/Properties/CoreStrings.resx [hunk 23]
```diff
@@ -1466,6 +1466,9 @@
   <data name="TempValuePersists" xml:space="preserve">
     <value>The property '{1_entityType}.{0_property}' has a temporary value while attempting to change the entity's state to '{state}'. Either set a permanent value explicitly, or ensure that the database is configured to generate values for this property.</value>
   </data>
+  <data name="TooManyArguments" xml:space="preserve">
+    <value>Too many arguments.</value>
```
Comment (AndriySvyryd) [inline src/EFCore/Properties/CoreStrings.resx:1470]: Add additional info, at least the method name

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 12 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:1066]: > Also make it static or inline

Can't make it static since it calls instance methods (e.g. Fragment). Can inline (which would duplicate), any particular reason to do that?

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 14 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:482]: > Where is the DecrementIndent?

The new GenerateRemainingAnnotations is responsible for both terminating the pre-existing chained call and decrementing the indent. This is necessary because we call it after already having indented (we're in the middle of a chained call), but it may need to generate type-qualified calls as well. This works the same in other places, added comments.

File: src/EFCore/Design/MethodCallCodeFragment.cs (hunk 16 above)
Comment (roji) [reply src/EFCore/Design/MethodCallCodeFragment.cs:146]: Agree, there are other ways we could tighten things, e.g. verify that the arguments are actually compatible with the parameters. At the generator level, we could also detect a differing return type and not chain on top of that.

Comment (roji) [reply src/EFCore/Design/MethodCallCodeFragment.cs:146]: #25498

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 15 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:1518]: Note also that I'm removing the old GenerateAnnotations and GenerateAnnotation, since they're dead-code.

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 18 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:1539]: There is an existing assumption that all method fragments coming from the provider's AnnotationCodeGenerator are either instance methods or extension methods (true static methods would have produced mangled code before this PR). Also, there's an assumption that all fragments can execute over the same builder - once again, a provider can't return calls which should happen on different builders (and that's not really possible to do since the builder identifier is handled externally by the snapshot generator).

So basically when we say "type-qualified", I think we're referring only to the very first call, to make sure we select the builder of the right provider (SQL Server vs. PostgreSQL). Once we have that builder (as it's returned by the first call), I don't see a problem with chaining over that...

What do you think?

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs [hunk 24]
```diff
@@ -110,7 +109,31 @@ public class CSharpDbContextGenerator : ICSharpDbContextGenerator
...
 
-            return _sb.ToString();
+            var namespaceStringBuilder = new StringBuilder();
+
+            IEnumerable<string> namespaces = _namespaces.OrderBy(
+                    ns => ns switch
+                    {
+                        "System" => 1,
+                        var s when s.StartsWith("System", StringComparison.CurrentCulture) => 2,
+                        var s when s.StartsWith("Microsoft", StringComparison.CurrentCulture) => 3,
```
[the remark this answers is not in the document]
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:119]: Yes, good catch, thanks.

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 19 above)
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:33]: Renamed to _builder, let me know if you want something else.

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 18 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:1505]: You're right, removing.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 12 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs]: Perf, but it doesn't matter that much

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 17 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs]: I wonder whether using `SortedDictionary` would be faster since they were already ordered initially.

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs (hunk 17 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs]: Interesting idea... Hard to know - we'd likely pay more in other accesses/lookups (e.g. when generating the fluent APIs, when removing ignorable/by-convention annotations...), who knows which way it goes.

I've seen lots of things around code generation here which aren't very optimized - that's probably OK. If we have indication that things are slow, at some point we can profile a bit and get rid of the worst offenders etc...

(BTW perf is also why I decided to discovery namespaces in scaffolding via a single pass along with generation, rather than the approach used in MigrationsCodeGenerator of a 2nd pass only for namespace discovery. Not sure how much all this is important...)

**Comment by roji** [comment]
Thanks @AndriySvyryd!
